hoomd/hpmc/ComputeSDF.h

```diff
@@ -364,13 +390,184 @@ template<class Shape> void ComputeSDF<Shape>::countHistogram(uint64_t timestep)
+            aabb.translate(pos_i_image);
+
+            // stackless search
+            for (unsigned int cur_node_idx = 0; cur_node_idx < aabb_tree.getNumNodes();
+                 cur_node_idx++)
+                {
+                if (detail::overlap(aabb_tree.getNodeAABB(cur_node_idx), aabb))
+                    {
+                    if (aabb_tree.isNodeLeaf(cur_node_idx))
+                        {
+                        for (unsigned int cur_p = 0;
+                             cur_p < aabb_tree.getNodeNumParticles(cur_node_idx);
+                             cur_p++)
+                            {
+                            // read in its position and orientation
+                            unsigned int j = aabb_tree.getNodeParticle(cur_node_idx, cur_p);
+
+                            // skip i==j in the 0 image
+                            if (cur_image == 0 && i == j)
+                                continue;
+
+                            Scalar4 postype_j = h_postype.data[j];
+                            Scalar4 orientation_j = h_orientation.data[j];
+                            int typ_j = __scalar_as_int(postype_j.w);
 
-        // record the minimum bin
-        if ((unsigned int)min_bin < m_hist.size())
-            m_hist[min_bin]++;
+                            // put particles in coordinate system of particle i
+                            vec3<Scalar> r_ij = vec3<Scalar>(postype_j) - pos_i_image;
 
+                            double u_ij_0 = 0.0; // energy of pair interaction in unperturbed state
+                            if (m_mc->m_patch)
+                                {
+                                u_ij_0 = m_mc->m_patch->energy(r_ij,
+                                                               typ_i,
+                                                               quat<float>(shape_i.orientation),
+                                                               float(h_diameter.data[i]),
+                                                               float(h_charge.data[i]),
+                                                               typ_j,
+                                                               quat<float>(orientation_j),
+                                                               float(h_diameter.data[j]),
+                                                               float(h_charge.data[j]));
+                                }
+                            // loop over bins; only going up to min_bin_ptl_i-1 since we only
+                            // want the _first_ overlap of particle i with its neighbors
+                            for (size_t bin_to_sample = 0; bin_to_sample < min_bin_ptl_i;
+                                 bin_to_sample++)
+                                {
+                                double scale_factor = m_dx * static_cast<double>(bin_to_sample + 1);
+
+                                // TODO: think about order of overlap checks: there may be a
```

Comment:
Have you tested this? Which way performs best?

Remove the TODO once you have chosen which order to evaluate the interactions.

Reply:
I have not tested this, but I think we do need to keep this order of overlap checks. This order ensures that if the scaling value results in both a new hard and soft overlap that the hard overlap (i.e., the one that dominates the Boltzmann weight) is the one that gets counted for the histogram. This scenario could may arise for systems with very short-ranged patchy interactions that are shorter-ranged than the compute's `dx`, but in that case the user should select SDF parameters that are more suited for the specific system.

Is this something worth looking more into?

Reply:
It is the user's responsibility to choose parameters.

Whichever order you choose to evaluate the overlaps, you should compute the same result. If there is a performance benefit to evaluating soft overlaps first, then you can use conditionals to add the correct value to the histogram.

In either case, I don't expect a performance benefit. In most loop iterations, you need to check for both hard and soft overlaps. Optimizations that evaluate only one overlap in the last loop iteration don't change that.